### DISCUSSION TITLE
No midi out port on windows?

### DISCUSSION THREAD

**Discussion by midenok** [FluidSynth/fluidsynth#1370] [Q&A]
Did I understand correctly fluidsynth does not create and listen on midi out via winmidi interface? It only listens on midi in of another devices. So I cannot play midi file in any directshow midi player to fluidsynth without setting up loopmidi (a bridge between midi out and midi in). Nobody nor documentation tells it explicitly like it is obvious, but to me it looks strange. Spent some time to figure that out myself.

**Comment by hikari-no-yume** [comment]
FluidSynth is primarily a library that can be used by other applications. There is a basic app built around it included with the project, but it is not very featureful. Appearing as its own MIDI device would be a fairly advanced feature. You might have more luck with one of the various apps that integrate FluidSynth, though I don't know myself if any have this feature: https://github.com/FluidSynth/fluidsynth/wiki/Applications

You might also be interested in “CoolSoft VirtualMIDISynth”, which definitely has this feature, but doesn't use FluidSynth.

**Reply by midenok** [reply]
Yes, I do use VirtualMIDISynth and loopMidi. As a user I don't understand the reason for such a limitation, therefore my perplexity. So it would be nice if this was explicit from the interface and/or documentation. Even Perplexity AI didn't know that (because no info anywhere on the net, I hope now there is because of this thread).

As a developer I can do a quick research and figure out it is because own driver is required and this project misses that, but that was not known from the start. I also see some projects on github supplying such kind of interfaces. I'm not really sure if there is one good enough.

**Comment by pedrolcl** [comment]
I can understand your frustration because Fluidsynth is not enough for your particular use case, but on the other hand please consider this other complementary scenario: A windows user buys an inexpensive MIDI controller like [this one](https://www.korg.com/products/computergear/nanokey2/) :

![MIDI Controller](https://m.media-amazon.com/images/I/51P9uj6yEJL.__AC_SY300_SX300_QL70_ML2_.jpg)

When the user tries to use it with the  windows native software synth or with VirtualMIDISynth, he will find that it is not possible. A virtual MIDI cable like loopMIDI does not help in this case either. Is this a limitation of the nanoKey2 or the synths?  Well, he can easily use this device in Linux and macOS with his synthesizer of choice, but not on Windows. I would say that this is a Windows limitation. Like in your use case.

And guess what: if the user installs Fluidsynth (or perhaps QSynth) on Windows, it will work with the same controller with little effort and without needing additional software.

To summarize:
* Windows does not provide virtual MIDI ports. This is the root of the limitation for your use case. Linux (ALSA) and macOS provide this feature natively, along with native MIDI connections. To overcome the limitation in Windows you need additional software: some virtual MIDI cable, like LoopMIDI. 
* Windows does not allow to connect two MIDI devices (two MIDI drivers) directly. In Linux and macOS all MIDI ports are equivalent, and any MIDI OUT can be connected to any MIDI IN using only commands and utilities from the operating system. This has been always a key feature of MIDI.

I've wrote about this [more than 10 years ago](https://midi-clorianos.blogspot.com/2011/04/midi-connection.html), but I'm not surprised that the artificial foolishness knows nothing about it.

**Reply by midenok** [reply]
Thanks for the link, that's useful to know. Yes, artificial foolishness cannot dig that deep. I guess if almost nobody needs the above scenario as there are better alternatives with DAWs.

**Reply by pedrolcl** [reply]
> there are better alternatives with DAWs.

Now that you mention DAWs: the limitations of Windows MIDI infrastructure is what created the necessity of [VSTi](https://en.wikipedia.org/wiki/Virtual_Studio_Technology) plugins in the first place (and a lucrative market for these type of products). You can find a [VSTi wrapper for fluidsynth here](https://github.com/prof-spock/FluidSynthPlugin) as well.

The plugin technology for MIDI synths was never indispensable in Linux or macOS, but the prevalence of Windows OS (and its comfort of using it) sparked the adoption of similar technologies like [Audio Units](https://en.wikipedia.org/wiki/Audio_Units) in Apple systems, and [LV2](https://en.wikipedia.org/wiki/LV2) on Linux.

And yes, you can use DAWs in Windows for musical MIDI production, thanks to plugins. What you can't do in Windows is to comfortably use MIDI software and hardware without DAWs and plugins.
